packages/theme/src/cli/utilities/theme-environment/theme-environment.ts

```diff
@@ -28,22 +29,27 @@ export function setupDevServer(theme: Theme, ctx: DevServerContext) {
 }
 
 function ensureThemeEnvironmentSetup(theme: Theme, ctx: DevServerContext) {
-  const remoteChecksumsPromise = fetchChecksums(theme.id, ctx.session)
-
-  const reconcilePromise = remoteChecksumsPromise.then((remoteChecksums) =>
-    handleThemeEditorSync(theme, ctx, remoteChecksums),
-  )
-
-  const uploadPromise = reconcilePromise.then(async ({updatedRemoteChecksumsPromise}) => {
-    const updatedRemoteChecksums = await updatedRemoteChecksumsPromise
-    return uploadTheme(theme, ctx.session, updatedRemoteChecksums, ctx.localThemeFileSystem, {
-      nodelete: ctx.options.noDelete,
-      deferPartialWork: true,
+  const abort = createAbortCatchError('Failed to perform the initial theme synchronization.')
+
+  const remoteChecksumsPromise = fetchChecksums(theme.id, ctx.session).catch(abort)
+
+  const reconcilePromise = remoteChecksumsPromise
+    .then((remoteChecksums) => handleThemeEditorSync(theme, ctx, remoteChecksums))
+    .catch(abort)
+
+  const uploadPromise = reconcilePromise
+    .then(async ({updatedRemoteChecksumsPromise}) => {
+      const updatedRemoteChecksums = await updatedRemoteChecksumsPromise
+      return uploadTheme(theme, ctx.session, updatedRemoteChecksums, ctx.localThemeFileSystem, {
+        nodelete: ctx.options.noDelete,
+        deferPartialWork: true,
+        backgroundWorkCatch: abort,
+      })
     })
-  })
+    .catch(abort)
```

Comment:
Instead of `.catch()`ing on each of these could we wrap the whole thing in a `try/catch`? Maybe not with the promises?

Reply:
It would be much more verbose with async promises than `.catch`, though. We could also pass it as the second param in `.then(() => {}, abort)` but I think most people are more familiar with `.catch` 🤔